From 4e3bcbb5f361fa8f5e9ae6aa81f4742d178b9e18 Mon Sep 17 00:00:00 2001
From: De Huo <De.Huo@windriver.com>
Date: Wed, 24 Jul 2019 10:28:57 +0800
Subject: [PATCH] Add rule accounts_passwords_pam_tally2_deny_root test case

Add rule accounts_passwords_pam_tally2_deny_root for WRLinux1019 product

Upstream-Status: Pending

Signed-off-by: De Huo <De.Huo@windriver.com>
---
 .../ansible/shared.yml                             | 81 ++++++++++++++++++++++
 .../bash/wrlinux1019.sh                            | 12 ++++
 .../oval/wrlinux1019.xml                           | 52 ++++++++++++++
 .../rule.yml                                       | 60 ++++++++++++++++
 4 files changed, 205 insertions(+)
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/ansible/shared.yml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/bash/wrlinux1019.sh
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/oval/wrlinux1019.xml
 create mode 100644 linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/rule.yml

diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/ansible/shared.yml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/ansible/shared.yml
new file mode 100644
index 0000000..d2b08c0
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/ansible/shared.yml
@@ -0,0 +1,81 @@
+# platform = multi_platform_rhel,multi_platform_fedora,multi_platform_ol,multi_platform_rhv
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
+- (xccdf-var var_accounts_passwords_pam_faillock_deny)
+
+- name: Add auth pam_faillock preauth deny before pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: sufficient
+    module_path: pam_unix.so
+    new_type: auth
+    new_control: required
+    new_module_path: pam_faillock.so
+    module_arguments: 'preauth
+        silent
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: before
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add deny argument to auth pam_faillock preauth
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: required
+    module_path: pam_faillock.so
+    module_arguments: 'preauth
+        silent
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: args_present
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add auth pam_faillock authfail deny after pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    control: sufficient
+    module_path: pam_unix.so
+    new_type: auth
+    new_control: '[default=die]'
+    new_module_path: pam_faillock.so
+    module_arguments: 'authfail
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: after
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add deny argument to auth pam_faillock authfail
+  pamd:
+    name: "{{ item }}"
+    type: auth
+    new_type: auth
+    control: '[default=die]'
+    module_path: pam_faillock.so
+    module_arguments: 'authfail
+        deny={{ var_accounts_passwords_pam_faillock_deny }}'
+    state: args_present
+  loop:
+    - system-auth
+    - password-auth
+
+- name: Add account pam_faillock before pam_unix.so
+  pamd:
+    name: "{{ item }}"
+    type: account
+    control: required
+    module_path: pam_unix.so
+    new_type: account
+    new_control: required
+    new_module_path: pam_faillock.so
+    state: before
+  loop:
+    - system-auth
+    - password-auth
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/bash/wrlinux1019.sh b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/bash/wrlinux1019.sh
new file mode 100644
index 0000000..1e23c88
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/bash/wrlinux1019.sh
@@ -0,0 +1,12 @@
+# platform = multi_platform_wrlinux
+
+. /usr/share/scap-security-guide/remediation_functions
+populate var_accounts_passwords_pam_tally2_deny
+
+if grep -q '^[^#].*pam_tally2.so' /etc/pam.d/common-auth; then
+  if ! grep -q 'root_unlock_time=' /etc/pam.d/common-auth; then
+    sed -i "/auth.*pam_tally2.so/s/$/ root_unlock_time=60/" /etc/pam.d/common-auth
+  fi
+else
+    echo "auth    required    pam_tally2.so deny=$var_accounts_passwords_pam_tally2_deny root_unlock_time=60" >>/etc/pam.d/common-auth
+fi
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/oval/wrlinux1019.xml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/oval/wrlinux1019.xml
new file mode 100644
index 0000000..a71e793
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/oval/wrlinux1019.xml
@@ -0,0 +1,52 @@
+<def-group>
+  <definition class="compliance" id="accounts_passwords_pam_tally2_deny_root" version="4">
+    <metadata>
+      <title>Lock out the root account after failed login attempts</title>
+      <affected family="unix">
+        <platform>multi_platform_wrlinux</platform>
+      </affected>
+      <description>The root account should be configured to deny access after the number of defined
+      failed attempts has been reached.</description>
+    </metadata>
+    <criteria operator="OR">
+      <criterion test_ref="test_accounts_passwords_pam_tally2_deny_root_unlock_time"
+      comment="pam_tally2.so root_unlock_time value set in common-auth" />
+
+      <criteria operator="AND" comment="root_unlock_time implies even_deny_root and fall back to unlock_time when absent">
+              <criterion test_ref="test_accounts_passwords_pam_tally2_deny_root" comment="pam_tally2.so even_deny_root set in common-auth" />
+              <criterion test_ref="test_accounts_passwords_pam_tally2_fall_back_unlock_time" comment="pam_tally2.so unlock_time set in common-auth" />
+      </criteria>
+    </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test id="test_accounts_passwords_pam_tally2_deny_root_unlock_time" check="all" check_existence="all_exist" comment="pam_tally2.so root_unlock_time value set in common-auth" version="1">
+    <ind:object object_ref="object_accounts_passwords_pam_tally2_deny_root_unlock_time" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_passwords_pam_tally2_deny_root_unlock_time" version="1">
+    <ind:filepath>/etc/pam.d/common-auth</ind:filepath>
+    <ind:pattern operation="pattern match">[\s]*auth[\s]+required[\s]+pam_tally2\.so[\s]+.*root_unlock_time=[0-9]+.*</ind:pattern>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <ind:textfilecontent54_test id="test_accounts_passwords_pam_tally2_deny_root" check="all" check_existence="all_exist" comment="pam_tally2.so even_deny_root set in common-auth" version="1">
+    <ind:object object_ref="object_accounts_passwords_pam_tally2_deny_root" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_passwords_pam_tally2_deny_root" version="1">
+    <ind:filepath>/etc/pam.d/common-auth</ind:filepath>
+    <ind:pattern operation="pattern match">[\s]*auth[\s]+required[\s]+pam_tally2\.so[\s]+.*even_deny_root.*</ind:pattern>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <ind:textfilecontent54_test id="test_accounts_passwords_pam_tally2_fall_back_unlock_time" check="all" check_existence="all_exist" comment="pam_tally2.so unlock_time set in common-auth" version="1">
+    <ind:object object_ref="object_accounts_passwords_pam_tally2_fall_back_unlock_time" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="object_accounts_passwords_pam_tally2_fall_back_unlock_time" version="2">
+    <ind:filepath>/etc/pam.d/common-auth</ind:filepath>
+    <ind:pattern operation="pattern match">^\s*auth\s+(?:(?:required))\s+pam_tally2\.so\s+.*\bunlock_time=([0-9]*).*$</ind:pattern>
+    <ind:instance datatype="int" operation="greater than or equal">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+</def-group>
diff --git a/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/rule.yml b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/rule.yml
new file mode 100644
index 0000000..29c18eb
--- /dev/null
+++ b/linux_os/guide/system/accounts/accounts-pam/locking_out_password_attempts/accounts_passwords_pam_tally2_deny_root/rule.yml
@@ -0,0 +1,60 @@
+documentation_complete: true
+
+prodtype: wrlinux1019,rhel6,rhel7,rhel8,fedora,ol7,ol8,rhv4
+
+title: 'Set Deny For Failed Password Attempts'
+
+description: |-
+    To configure the system to lock out accounts after a number of incorrect login
+    attempts using <tt>pam_tally2.so</tt>, modify the content of both
+    <tt>/etc/pam.d/common-auth</tt> and <tt>/etc/pam.d/common-account</tt> as follows:
+    <br /><br />
+    <ul>
+    <li> add the following line immediately <tt>before</tt> the <tt>pam_unix.so</tt> statement in the <tt>AUTH</tt> section:
+    <pre>auth required pam_tally2.so preauth silent deny=<sub idref="var_accounts_passwords_pam_tally2_deny" /> unlock_time=<sub idref="var_accounts_passwords_pam_tally2_deny" /> fail_interval=<sub idref="var_accounts_passwords_pam_faillock_fail_interval" /></pre></li>
+    <li> add the following line immediately <tt>after</tt> the <tt>pam_unix.so</tt> statement in the <tt>AUTH</tt> section:
+    <pre>auth [default=die] pam_tally2.so authfail deny=<sub idref="var_accounts_passwords_pam_tally2_deny" /> unlock_time=<sub idref="var_accounts_passwords_pam_faillock_unlock_time" /> fail_interval=<sub idref="var_accounts_passwords_pam_faillock_fail_interval" /></pre></li>
+    <li> add the following line immediately <tt>before</tt> the <tt>pam_unix.so</tt> statement in the <tt>ACCOUNT</tt> section:
+    <pre>account required pam_tally2.so</pre></li>
+    </ul>
+
+rationale: |-
+    Locking out user accounts after a number of incorrect attempts
+    prevents direct password guessing attacks.
+
+severity: medium
+
+identifiers:
+    cce@rhel6: 26844-1
+    cce@rhel7: 27350-8
+    cce@rhel8: 80667-9
+
+references:
+    stigid@rhel6: RHEL-06-000061
+    srg@rhel6: SRG-OS-000021
+    disa@rhel6: '44'
+    cis: 5.3.2
+    cjis: 5.5.3
+    cui: 3.1.8
+    disa: "2238,44"
+    nist: AC-7(a)
+    nist-csf: PR.AC-7
+    ospp: FMT_MOF_EXT.1
+    pcidss: Req-8.1.6
+    srg: SRG-OS-000329-GPOS-00128,SRG-OS-000021-GPOS-00005
+    vmmsrg: SRG-OS-000021-VMM-000050
+    stigid@rhel7: "010320"
+    isa-62443-2013: 'SR 1.1,SR 1.10,SR 1.2,SR 1.5,SR 1.7,SR 1.8,SR 1.9'
+    isa-62443-2009: 4.3.3.6.1,4.3.3.6.2,4.3.3.6.3,4.3.3.6.4,4.3.3.6.5,4.3.3.6.6,4.3.3.6.7,4.3.3.6.8,4.3.3.6.9
+    cobit5: DSS05.04,DSS05.10,DSS06.10
+    iso27001-2013: A.18.1.4,A.9.2.1,A.9.2.4,A.9.3.1,A.9.4.2,A.9.4.3
+    cis-csc: 1,12,15,16
+
+ocil_clause: 'that is not the case'
+
+ocil: |-
+    To ensure the failed password attempt policy is configured correctly, run the following command:
+    <pre>$ grep pam_tally2 /etc/pam.d/system-auth</pre>
+    The output should show <tt>deny=<sub idref="var_accounts_passwords_pam_tally2_deny" /></tt>.
+
+platform: pam
-- 
1.9.1

